rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -328,6 +350,134 @@ bool Player::play_next()
+  rosbag2_storage::SerializedBagMessageSharedPtr message_before_seek;
+  if (message_ptr != nullptr) {
+    message_before_seek = *message_ptr;
+  }
+
+  {  // Update message queue and adjust reader pointer
+    std::lock_guard<std::mutex> lk(reader_mutex_);
+    // Read next current_message
+    rosbag2_storage::SerializedBagMessageSharedPtr current_message;
+    if (!message_queue_.try_dequeue(current_message)) {
+      if (!reader_->has_next()) {
+        reader_->close();
+        reader_->open(storage_options_, {"", rmw_get_serialization_format()});
+        if (!reader_->has_next()) {
+          return false;  // The bag is empty
+        }
+        // else means that we were at the end of the bag when we were started and will try to
+        // search from the beginning
+        assert(is_storage_completely_loaded());
+        storage_loading_future_ =
+          std::async(std::launch::async, [this]() {load_storage_content();});
+      }
+      current_message = reader_->read_next();
+    }
+
+    const bool jump_forward = (time_point >= current_message->time_stamp);
+
+    if (jump_forward) {
+      if (current_message->time_stamp >= time_point) {
+        message_found = true;
+      }
+      // Pop up current_message queue trying to find current_message with needed timestamp
+      while (!message_found && message_queue_.try_dequeue(current_message)) {
+        if (current_message->time_stamp >= time_point) {
+          message_found = true;
+        }
+      }
+      while (!message_found && reader_->has_next()) {
+        current_message = reader_->read_next();
+        if (current_message->time_stamp >= time_point) {
+          message_found = true;
+        }
+      }
+      if (message_found) {
+        // Enqueue current_message and rest of the queue to the queue again to preserve order of
+        // messages
+        if (message_queue_.peek() != nullptr) {  // if message_queue_ not empty
+          decltype(message_queue_) temp_message_queue;
+          temp_message_queue.enqueue(current_message);
+          // Copy residual messages from message_queue_ to the temp_message_queue
+          while (message_queue_.try_dequeue(current_message)) {
+            temp_message_queue.enqueue(current_message);
+          }
+          // Copy all messages back to the message_queue_
+          while (temp_message_queue.try_dequeue(current_message)) {
+            message_queue_.enqueue(current_message);
+          }
+          enqueue_up_to_boundary(play_options_.read_ahead_queue_size);  // refill queue
+        } else {
+          message_queue_.enqueue(current_message);
+          enqueue_up_to_boundary(play_options_.read_ahead_queue_size);  // refill queue
+        }
+      } else {
+        restore_message_queue_from_storage(message_before_seek);
+      }
+    } else {  // jump_backward
+      // Purge queue
+      while (message_queue_.pop()) {}
```

Comment:
would it make more sense to reconstruct the queue rather than pop each element individually?

Reply:
But how to properly reconstruct the `message_queue_` if we are allocating it on the stack?
I can try to use a "hack" with move constructor or move assignment operator. But it will be implementation specific and might not properly work in future, since in general case object shouldn't be used after move.